core/unix/signal.c

```diff
@@ -6693,6 +6700,141 @@ execute_native_handler(dcontext_t *dcontext, int sig, sigframe_rt_t *our_frame)
+    //
+    // Summary of relevant events when DR starts detaching:
+    // D: Detacher thread sends suspend signal to all other threads
+    // O: each of the Other threads receive the suspend signal. Note that DR's
+    //    signal handling configuration blocks all signals except SIGSEGV and
+    //    the suspend signal. So at this point Other threads cannot receive
+    //    app signals. Other threads signal to the Detacher thread that they're
+    //    suspended and awaiting wakeup call to continue detaching (sig_detach).
+    // D: Detacher thread sets doing_detach to true. A side effect of this is
+    //    that get_thread_private_dcontext() will return NULL when any Other
+    //    thread tries to get its private dcontext.
+    // D: Detacher thread wakes up each Other thread, telling it to detach.
+    // O: Each Other thread wakes up, does sig_detach which reinstates the app
+    //    signal stack (if available) and the app's blocked signal set, lets the
+    //    Detacher thread know that its done detaching and resumes native
+    //    execution. Note that at this point, DR's sigact config is still
+    //    installed (including main_signal_handler). When the Other thread
+    //    return from the suspend signal, signals get unblocked automatically
+    //    and may now be delivered.
+    // <execute_native_handler_using_cur_frame may help here>
+    // D: When all threads are done detaching, the Detacher thread cleans up
+    //    each thread's drcontext for real, and also reinstates the app's
+    //    sigact in signal_thread_exit (which removes main_signal_handler).
+    //
+    // Therefore, if we arrive here and find doing_detach to be true and
+    // dcontext to be NULL, we can be certain that the current thread has
+    // completed sig_detach. This means the interrupted context stack (sc->XSP)
+    // must be the app's own regular stack.
+    //
+    // Also, if the thread has a signal stack at all for any signal, we will
+    // be on it (because DR sets SA_ONSTACK for all signals, and that config is
+    // still in effect), regardless of whether the signal stack was registered
+    // for sig. If the signal stack was actually not registered for sig, then we
+    // return true (indicating that DR should continue trying to deliver the
+    // signal).
+    if (dcontext != NULL || !doing_detach)
+        return false;
+
+    bool has_sigstack = false;
+#ifdef HAVE_SIGALTSTACK
+    stack_t app_sigstack;
+    IF_DEBUG(int rc =)
+    sigaltstack_syscall(NULL, &app_sigstack);
+    ASSERT(rc == 0);
+    if (APP_HAS_SIGSTACK_EX(app_sigstack)) {
+        has_sigstack = true;
+        // If the app has a sigstack, we must be on it.
+        ASSERT(app_sigstack.ss_sp < (void *)xsp &&
+               (void *)xsp < app_sigstack.ss_sp + app_sigstack.ss_size);
+        // In release build, we bail on this optimization and try to deliver the
+        // signal.
+        return false;
```

Comment:
Why do we give up if the app has an altstack?

Reply:
We want to give up if there's an altstack but we're not on it. Fixed it.